### ISSUE TITLE
Proposal for a “Delta Compression + Neural Network” Approach in Nine Men’s Morris

### ISSUE THREAD

**Issue by calcitem** [ggevay/malom#4]
Dear Mr. Gévay and Mr. Danner,

I hope this message finds you well. I have been greatly inspired by your work on Nine Men’s Morris, Morabaraba, and Lasker Morris, particularly regarding extended and ultrastrong solutions. Building on the foundations you established, I am developing a preliminary research plan to integrate **perfect database knowledge** with **traditional Alpha-Beta search** in a more streamlined way, ultimately aiming to train a **lightweight neural network** for efficient, near-optimal play.

Below is the **detailed design** of my intended approach. While none of these steps have been implemented yet, I would be grateful for any guidance or suggestions you might share.

---

## 1. Overall Goal

Create a **“delta-compressed”** subset of the tablebase that captures only those states in which Alpha-Beta search fails to replicate the optimal solution. This compressed database should be dramatically smaller yet still ensure overall near-perfect decision-making.

---

### 2.1 Segmenting the Database

1. **Phased or Subset Handling:**  
   - We plan to split the full tablebase by game phase (e.g., opening/placing, midgame/moving, endgame/flying) or other meaningful partitions (piece counts, forced lines, etc.).  
   - By working on smaller subsets, we can systematically test and refine our methods before extending them to the entire database.

2. **Sampling Strategies:**  
   - For each subset, we may combine **random sampling** for broad coverage and **stratified sampling** to focus on higher-complexity states.  
   - This ensures we do not overlook “rare but critical” positions.

### 2.2 Alpha-Beta Search Validation

1. **Running Alpha-Beta:**  
   - We will use a minimax framework with iterative deepening, heuristic evaluation, and basic pruning.  
   - The search depth or time budget can be adjusted per subset to balance thoroughness and computation.

2. **Mismatch Detection:**  
   - For every selected state \(s\), we compare the Alpha-Beta best move (and outcome, if applicable) with the tablebase result.  
   - If Alpha-Beta **matches** the database, we need not store additional information.  
   - If Alpha-Beta **differs**, we mark \(s\) as a conflict and record essential data (optimal move, value, or short continuation).

### 2.3 Building the “Delta” Compressed Database

1. **Core Principle:**  
   - Only states where Alpha-Beta conflicts with the tablebase are retained in the “delta” set.  
   - Each entry contains:  
     - **State ID** (e.g., Zobrist hash or another representation).  
     - **Correct move/value** according to the perfect database.  
     - **Optional follow-up** for forced lines, ensuring correct transitions in deeper sequences.

2. **Iterative Construction:**  
   - We plan to add entries to the delta set incrementally, continuing until the main search engine rarely encounters undiscovered discrepancies.

3. **Use in Gameplay:**  
   - During a live match, the system would run Alpha-Beta . For state \(s\):  
     - If \(s\) is **not** in the delta set, Alpha-Beta is presumed correct.  
     - If \(s\) **is** in the delta set, we override Alpha-Beta with the tablebase’s recommended move.

### 2.4 Neural Network Training

1. **Data Sources:**  
   - **Consistent States:** (where Alpha-Beta aligns with the tablebase) used to train on “typical” or “straightforward” positions.  
   - **Conflict States:** (from the delta table) used to highlight trickier scenarios in which Alpha-Beta needed correction.

2. **Network Structure:**  
   - Potentially a convolutional or graph-based architecture to encode Nine Men’s Morris boards (24 points, occupant info, etc.).  
   - Outputs could be a **value** (win/draw/loss or numerical) and/or a **policy** (move probabilities).

3. **Training Workflow:**  
   - **Supervised Phase:** We label each sampled position with the tablebase’s best move or value.  
   - **Iterative Refinement:** If the network remains uncertain in certain sub-regions, we may sample more states or refine heuristics for those areas, always referencing the delta table for ground truth.

---

## 3. Potential Refinements

1. **Symmetry and Transformation:**  
   - Exploit rotational/reflective symmetries to reduce storage and training data repetition.  
2. **Further Distillation & Self-Play:**  
   - After the initial supervised stage, it could be beneficial to integrate minimal self-play or a reinforcement step, referencing the delta table whenever the network encounters new or uncertain positions.  
3. **Coverage vs. Table Size:**  
   - We aim to keep the compressed database small, but must ensure that all critical lines—especially forced sequences—are captured.  
4. **Practical Heuristics:**  
   - If certain states are extremely deep or complex, we plan to refine our Alpha-Beta heuristics or incrementally increase search depth to catch hidden traps.

---

## 4. Invitation for Your Guidance

As we have not yet begun implementation, your expertise could help us refine the initial strategy and prioritize our efforts. We welcome any broad or specific suggestions, such as:

- **Ensuring comprehensive coverage** of intricate states without over-expanding the delta table.  
- **Efficient indexing** (e.g., hashing strategies) to guarantee fast lookups in both the tablebase and the compressed data.  
- **Integrating knowledge of specialized lines** (e.g., punishing suboptimal moves) in the style of ultrastrong solutions.  

We would be honored to incorporate any insights or recommendations you might share. 

Thank you for taking the time to review this proposal. Your research has been a significant inspiration, and I look forward to any thoughts or advice you may be willing to offer.

Warm regards,

Calcitem

**Comment by ggevay** [comment]
Hello!

These are interesting ideas!

One potential difficulty that I can see is that if there is a time limit for the Alpha-Beta search, then it's not deterministic: even if it finds an optimal move in a certain game state on your machine, maybe it doesn't find an optimal move in the same game state on a user's machine. To circumvent this problem, you could use a node limit instead of a time limit: during searches, you could keep track of the number of nodes explored in the game tree, and stop the iterative deepening when this number reaches a limit. This has the drawback that the program will spend different times on different computers, but it might be worth it. You might be able to set the limit low enough that it plays reasonably fast on most computers.

Another difficulty could be finding a _representative_ sample of important game states: If you play games with your program, it might not wander into many game states that human users or other programs would wander into. One thing you could do is to play games with many different engine configurations, and maybe even with different tweaks to the evaluation function.

In win/loss positions, will you consider optimality with taking into account the number of moves to the end of the game (depth-to-win)? If yes, then the Alpha-Beta search will have a tougher time to find an optimal move: finding the quickest way to win is often harder than just finding any win. This means that not taking into account depth-to-win could potentially make the delta database much smaller. However, not taking into account depth-to-win (i.e., just going with any winning move), can cause a program to get into a cycle where every position of the cycle is a winning position, but the program is not actually winning, because of being in a cycle. A middle ground between not taking into account depth-to-win at all and taking it into account at every win/loss position could be to just pick one position of a cycle of winning positions that the Alpha-Beta runs into, and include just that one position in the delta database, so that the cycle is broken. This will prevent the program from running around in such cycles instead of winning, but at the same time would make the database much smaller than simply taking into account the depth-to-win when determining optimality in all win/loss positions.

I'm curious to see how your research turns out! It will be interesting to see statistics about how often the Alpha-Beta search can find an optimal move.

One more thing: if the delta database fits in RAM, you could use it not just to look up the value of the game state that is on the board (i.e., the root of a potential Alpha-Beta search), but to look up every game state that you encounter during the Alpha-Beta search, and then not explore it with the Alpha-Beta search if it's present in the delta database as a win/loss (if it's just present in the database as a draw, then it's less useful for the Alpha-Beta search; in that case, it's just a very loose bound on the result that the Alpha-Beta search would return: it's somewhere between the best loss and the worst win). It seems hard to estimate how big of a boost this would give to the Alpha-Beta search, but it seems like an interesting experiment at the very least.

**Comment by calcitem** [comment]
Hello!

Thank you again for sharing your suggestions! Below is a more in-depth explanation of how we plan to handle some of the issues you raised—particularly regarding deterministic search depth (rather than time or node limits), the potential impact of 50-move/threefold repetition, and overall considerations for integrating the delta-based table with a traditional Alpha-Beta framework.

---

## 1. Fixed-Depth vs. Time/Node Constraints

Instead of using a time limit or a node limit, we intend to cap the search at a fixed depth (e.g., 15 plies). This makes the engine’s behavior fully deterministic—every run will explore the same search tree depth regardless of CPU differences. As you pointed out, using a time or node budget often leads to non-deterministic outcomes, which can cause inconsistencies in whether a position is recognized as winning/losing by Alpha-Beta.

### 1.1 Iterative Deepening for Manageability

Although we set a final target depth (e.g., 15 plies), the search procedure still employs **iterative deepening** from 1 up to 15. This allows partial solutions at intermediate depths to be updated incrementally. We can stop early in practice if certain draw or repetition conditions appear (see the next section) or if we detect a forced mate. The relevant code is in `SearchEngine::executeSearch()`:

```cpp
for (Depth i = depthBegin; i < originDepth; i += 1) {
    // ...
    value = Search::search(*this, rootPos, ss, i, i, alpha, beta, bestMove);
    // or MTD(f), or MCTS
    // ...
}
```

Since we do not rely on any absolute “time out” or “node out,” the engine is consistent across different machines or different user configurations.

---

## 2. Handling 50-Move Rule and Threefold Repetition

You also mentioned the possibility of “50-move rule” or being stuck in a cycle (threefold repetition). Our engine includes logic to detect these scenarios so we can properly terminate or declare a draw, preventing infinite searching or “threefold blindness.”

### 2.1 50-Move Rule

Inside `search()` and `executeSearch()`, you will see conditional blocks for `#ifdef RULE_50` and references to `pos->rule50_count()`. Essentially, we keep a move counter (`posKeyHistory.size()`) that increments each time a non-capturing move is made. If this counter exceeds a threshold such as 50 moves—or a specialized rule in an endgame scenario—Alpha-Beta will detect a “draw by rule.” That triggers an early cutoff:

```cpp
if (pos->rule50_count() > rule.nMoveRule
    || (rule.endgameNMoveRule < rule.nMoveRule && pos->is_three_endgame()
        && pos->rule50_count() >= rule.endgameNMoveRule))
{
    alpha = VALUE_DRAW;
    if (alpha >= beta) {
        return alpha; 
    }
}
```

So, even at fixed depth, we won’t get stuck in a position that’s effectively a draw by the 50-move rule.

### 2.2 Threefold Repetition

We also track position keys (`pos->has_repeated(ss)`) to detect repeated board states. In Nine Men’s Morris variants, this can happen if both sides keep rotating among the same positions. When repetition is detected:

```cpp
if (depth != originDepth && pos->has_repeated(ss)) {
    alpha = VALUE_DRAW;
    if (alpha >= beta) {
        return alpha;
    }
}
```

Thus, the search short-circuits the evaluation to a draw, ensuring we do not explore repetitive sequences forever.

---

## 3. Practical Implications for the Delta Database

As for merging a perfect tablebase with Alpha-Beta:

1. **Fixed Depth Determinism**: By searching to a strict depth, any discrepancy we identify for the “delta” table can be reliably reproduced. We don’t risk a user’s hardware suddenly failing to reach the needed depth due to time constraints.

2. **Dealing with Depth-to-Win**: We currently prefer including depth information in the evaluation (e.g., `stand_pat += depth` for positive scores) to encourage faster mates. However, we agree there can be an exponential blow-up in the number of positions if we store full “optimal to the move count” solutions. Your idea to skip full depth-to-win tracking for every position, and instead break cycles by only tagging a single “cycle state” in the database, is a great middle ground. We may incorporate that to keep the delta set compact.

3. **Coverage of “Critical” States**: Because some states are rarely visited in normal self-play (or differ from “human-style” play), we plan to sample from multiple engine configurations, random or pseudo-random moves, and known opening lines to find tricky positions. This expands coverage for the delta database beyond the typical search tree.

4. **Future Search Enhancement**: If the delta table fits in RAM, we can indeed do a quick check in the transposition table or a specialized hash map at each node of the Alpha-Beta search. If a position is recognized as a forced win/loss in the delta table, we can immediately prune that branch. We have hooks for transposition table lookups (`TranspositionTable::probe()` in `search.cpp`) where we could easily incorporate a delta table lookup for faster detection of known states.

---

## 4. Conclusion and Next Steps

1. **Fixed Depth**: Our decision is to fix the search to (say) 15 plies for determinism, while also employing standard iterative deepening. The 50-move rule and threefold detection code will ensure no infinite loops.
2. **Delta Table Minimization**: We are keeping the focus on “conflict positions” where Alpha-Beta deviates from perfect data, but we are open to skipping precise depth-to-win for all states.  
3. **Integration**: Once the conflict set is stable, we will integrate it into the transposition table and also train a small neural network. The code references you see in `search.cpp` (`TT.clear()`, etc.) can be adapted for injecting delta-based knowledge.

We greatly appreciate your input—your ideas about simplifying depth-to-win cycles and sampling from various engine configurations will help us refine both our table building and our training data. If you have any further insights or concerns, especially regarding memory constraints or coverage for rarely occurring but critical board positions, we would love to hear them.

Many thanks again, and we look forward to sharing new results as the implementation progresses!

**Comment by ggevay** [comment]
Sounds good!

**Comment by calcitem** [comment]
Hello! I've written an initial version of a program for stripping the database:

[perfect_strip_db.cpp](https://github.com/calcitem/Sanmill/blob/perfect/src/perfect/perfect_strip_db.cpp)

I haven't tested it yet, and I'm not very familiar with the structure of the DB files. If you have some time, would you mind checking whether the file parsing and read/write sections are correct?

Thank you very much!
